src/globus_sdk/experimental/globus_app/globus_app.py

```diff
@@ -0,0 +1,429 @@
+            if auth_params.required_scopes:
+                combined_scopes = Scope.merge_scopes(
+                    required_scopes, [Scope(s) for s in auth_params.required_scopes]
+                )
+                auth_params.required_scopes = [str(s) for s in combined_scopes]
+            else:
+                auth_params.required_scopes = [str(s) for s in required_scopes]
+            return auth_params
+
+        else:
+            return GlobusAuthorizationParameters(
+                required_scopes=[str(s) for s in required_scopes]
+            )
+
+    def get_authorizer(self, resource_server: str) -> GlobusAuthorizer:
+        """
+        Get a ``GlobusAuthorizer`` from the app's authorizer factory for a specified
+        resource server. The type of authorizer is dependent on the app.
+
+        :param resource_server: the resource server the Authorizer will provide
+            authorization headers for
+        """
+        return self._authorizer_factory.get_authorizer(resource_server)
+
+    def add_scope_requirements(
+        self, scope_requirements: dict[str, list[Scope]]
+    ) -> None:
+        """
+        Add given scope requirements to the app's scope requirements.
+
+        :param scope_requirements: a dict of Scopes indexed by resource server
+            that will be added to this app's scope requirements
+        """
+        for resource_server, scopes in scope_requirements.items():
+            if resource_server not in self._scope_requirements:
+                self._scope_requirements[resource_server] = scopes
+
+            else:
+                self._scope_requirements[resource_server] = Scope.merge_scopes(
+                    self._scope_requirements[resource_server], scopes
+                )
+
+        # update the app's ValidatingTokenStorage's scope requirements to enforce the
+        # updated scope requirements
+        self._validating_token_storage.scope_requirements = self._scope_requirements
+
+
+class UserApp(GlobusApp):
+    """
+    A ``GlobusApp`` for login methods that require an interactive flow with a user.
+
+    Can either use a native application client by only giving ``client_id`` or a
+    templated client by giving both ``client_id`` and ``client_secret``.
+
+    .. tab-set::
+
+        .. tab-item:: Future Example Usage
+
+            .. code-block:: python
+
+                app = UserApp("myapp", client_id=NATIVE_APP_CLIENT_ID)
+                client = TransferClient(app=app)
+                res = client.operation_ls(COLLECTION_ID)
+
+    """
+
+    _login_client: NativeAppAuthClient | ConfidentialAppAuthClient
+    _authorizer_factory: (  # type:ignore
+        AccessTokenAuthorizerFactory | RefreshTokenAuthorizerFactory
+    )
+
+    def __init__(
+        self,
+        app_name: str,
+        *,
+        scope_requirements: dict[str, list[Scope]] | None = None,
+        client_id: UUIDLike | None = None,
+        client_secret: str | None = None,
```

Comment:
We should also accept a pre-instantiated login client here (mutexed with these client id/secret values) to make it easier for users to plug this into their existing applications that either (1) re-use the login client in multiple places or (2) extend and customize a particular login client.

(Comment applies to ClientApp as well)